src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -601,32 +895,119 @@ private static PartialCaseDecision RemoveEvaluation(PartialCaseDecision c, Bound
                 Index: c.Index,
                 Syntax: c.Syntax,
                 Decisions: c.Decisions.WhereAsArray(d => !(d is BoundDagEvaluation e2) || e2 != e),
-                Bindings: c.Bindings, WhereClause: c.WhereClause, CaseLabel: c.CaseLabel);
+                Bindings: c.Bindings, WhenClause: c.WhenClause, CaseLabel: c.CaseLabel);
+        }
+
+        private class DagState
+        {
+            public readonly ImmutableArray<PartialCaseDecision> Cases;
+
+            public DagState(ImmutableArray<PartialCaseDecision> cases)
+            {
+                this.Cases = cases;
+            }
+
+            public static implicit operator DagState(ImmutableArray<PartialCaseDecision> cases) => new DagState(cases);
+
+            // We only compute the dag states for the branches after we de-dup this DagState itself.
+            // If all that remains is the `when` clauses, SelectedDecision is left `null` and the
+            // FalseBranch field is populated with the successor on failure of the when clause (if one exists).
+            public BoundDagDecision SelectedDecision;
+            public DagState TrueBranch, FalseBranch;
+
+            // After the entire graph of DagState objects is complete, we translate each into its Dag.
+            public BoundDecisionDag Dag;
         }
 
-        internal class PartialCaseDecision
+        /// <summary>
+        /// An equivalence relation between dag states used to dedup the states during dag construction.
+        /// After dag construction is complete we treat a DagState as using object equality as equivalent
+        /// states have been merged.
+        /// </summary>
+        private class DagStateEquivalence : IEqualityComparer<DagState>
         {
+            public static readonly DagStateEquivalence Instance = new DagStateEquivalence();
```

Comment:
Any reason not to add a private constructor?

Reply:
Because the compiler provides one.

---
In reply to: [176222190](https://github.com/dotnet/roslyn/pull/24059#discussion_r176222190) [](ancestors = 176222190)

Reply:
Oh, I see, to prevent it from being constructed elsewhere. Recorded in [https://github.com/dotnet/roslyn/issues/25663](https://github.com/dotnet/roslyn/issues/25663)

---
In reply to: [176307206](https://github.com/dotnet/roslyn/pull/24059#discussion_r176307206) [](ancestors = 176307206,176222190)